sdk/src/Core/Amazon.Util/AWSSDKUtils.cs

```diff
@@ -478,6 +501,25 @@ public static string JoinResourcePathSegmentsV2(IEnumerable<string> pathSegments
             return string.Join(Slash, pathSegments.ToArray());
         }
 
+        /// <summary>
+        /// Joins all path segments with the / character and encodes each non-literal segment before joining
+        /// </summary>
+        /// <param name="pathSegments"></param>
+        /// <returns></returns>
+        public static string JoinResourcePathSegmentsV2(IEnumerable<UriComponent> pathSegments)
+        {
+            List<string> encodedSegments = pathSegments.Select(segment =>
+            {
+                if (segment.SegmentType == SegmentType.Label)
+                    return UrlEncode(segment.Value, false);
+                else
+                    return UrlEncode(segment.Value, true);
+            }).ToList();
+            // join the encoded segments with /
+            return string.Join(Slash, encodedSegments.ToArray());
```

Comment:
Join has enumerable support. Why do you need to eager materialize twice?

Reply:
Ah, you're right I can get rid of the second `.ToArray()`